### DISCUSSION TITLE
Facilitate passing custom CA cert into sidecar containers

### DISCUSSION THREAD

**Discussion by hrobertson** [renovatebot/renovate#24836] [Suggest an Idea]
### What would you like Renovate to be able to do?

We have various self-hosted infrastructure such as git (GitLab) and package repositories (Artifactory) which have certificates issued by our own CA.
We need to use `binarySource=docker` (`renovate:slim`) as we have projects using various versions of package managers.
We can set `NODE_EXTRA_CA_CERTS` within the renovate container, but another solution is needed to enable the sidecars to trust our CA.

One option is for us to create our own sidecar images that include the CA cert ( as recommended by https://github.com/renovatebot/renovate/issues/2835#issuecomment-725834203) but this does not seem like an ideal solution.

A preferable solution would be to pass the CA certificate into the sidecars. This would merely involve configuring the core renovate appropriately rather than maintaining a load of sidecar images.

### If you have any ideas on how this should be implemented, please tell us here.

**Summary**
Mount a volume containing the certificate in the sidecar, and run a command or set an environment variable as appropriate to make use of the cert. See https://github.com/containerbase/base/blob/main/docs/custom-root-ca.md

**Detail**
A bot administrator would be able to use a global config variable (a la `dockerUser`, eg `dockerCACert`) to specify the location of the CA cert to be mounted. [generateDockerCommand](https://github.com/renovatebot/renovate/blob/main/lib/util/exec/docker/index.ts#L211) would mount this to a well known mount point.

Different tools have different requirements - some only require an environment variable (`SSL_CERT_FILE` or in the case of node `NODE_EXTRA_CA_CERTS`), others require the cert be installed with `update-ca-certificates`. It would therefore make sense for each manager to specify the `extraEnv` or `preCommand` appropriate for it when calling [exec](https://github.com/renovatebot/renovate/blob/main/lib/util/exec/index.ts#L168). It would only add this if the global config `dockerCACert` was set.

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by rarkins** [comment]
Have you looked into binarySource=install lately, and not using sidecars? It's hopefully a viable alternative

**Comment by hrobertson** [comment]
Thanks, I hadn't considered that. I'll give it a go.

**Comment by lovetheguitar** [comment]
With `RENOVATE_BINARY_SOURCE=install` I get
```console
installing v2 tool python v3.11.1
python not prepared
```
(similar to https://github.com/renovatebot/renovate/discussions/16810#discussioncomment-4125016)

Any other solution/workaround than building ourselves?

**Comment by viceice** [comment]
you currently need to use the full image for install mode.

**Comment by rarkins** [comment]
Or run as root? :-/ (usually not recommended)

**Comment by lovetheguitar** [comment]
> you currently need to use the full image for install mode.

Thanks for the fast reply. Changed to that and now poetry installs correctly.

I set up `REQUESTS_CA_BUNDLE` pointing to a Gitlab CI file variable but I still get `CERTIFICATE_VERIFY_FAILED`
<details>
<summary>Detailed Output</summary>

```console
DEBUG: Executing command (repository=hw/some_other_lib, branch=renovate/some-lib-0.x)
       "command": "poetry update --lock --no-interaction some-lib"
DEBUG: rawExec err (repository=hw/some_other_lib, branch=renovate/some-lib-0.x)
       "err": {
         "name": "ExecError",
         "cmd": "/bin/sh -c poetry update --lock --no-interaction some-lib",
         "stderr": "Creating virtualenv ´some-other-lib-yjxm3t6G-py3.11 in /home/ubuntu/.cache/pypoetry/virtualenvs\n\nHTTPSConnectionPool(host='pypi.internal-domain.com', port=443): Max retries exceeded with url: /simple/some-lib/ (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:992)')))\n",
         "stdout": "Updating dependencies\nResolving dependencies...\n",
         "options": {
           "cwd": "/builds/hw/renovate_bot/renovate/repos/gitlab/hw/some_other_lib",
           "encoding": "utf-8",
           "env": {
             "PIP_CACHE_DIR": "/builds/hw/renovate_bot/renovate/cache/others/pip",
             "HOME": "/home/ubuntu",
             "PATH": "/home/ubuntu/.cargo/bin:/home/ubuntu/.local/bin:/go/bin:/home/ubuntu/bin:/opt/buildpack/tools/python/3.11.1/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/.cargo/bin:/home/ubuntu/.local/bin:/go/bin:/home/ubuntu/bin:/opt/buildpack/tools/python/3.11.1/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/.cargo/bin:/home/ubuntu/.local/bin:/go/bin:/home/ubuntu/bin:/opt/buildpack/tools/python/3.11.1/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
             "LC_ALL": "C.UTF-8",
             "LANG": "C.UTF-8",
             "DOCKER_HOST": "tcp://docker:2376",
             "DOCKER_TLS_VERIFY": "1",
             "DOCKER_CERT_PATH": "/certs/client",
             "BUILDPACK_CACHE_DIR": "/builds/hw/renovate_bot/renovate/cache/containerbase",
             "CONTAINERBASE_CACHE_DIR": "/builds/hw/renovate_bot/renovate/cache/containerbase"
           },
           "maxBuffer": 10485760,
           "timeout": 900000
         },
         "exitCode": 1,
         "message": "Command failed: poetry update --lock --no-interaction some-lib\nCreating virtualenv ´some-other-lib-yjxm3t6G-py3.11 in /home/ubuntu/.cache/pypoetry/virtualenvs\n\nHTTPSConnectionPool(host='pypi.internal-domain.com', port=443): Max retries exceeded with url: /simple/some-lib/ (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:992)')))\n",
```
</details>

**Update**
Adding this to my `config.js` made poetry work. 🥳 
```js
"customEnvVariables": {
    "REQUESTS_CA_BUNDLE": process.env.REQUESTS_CA_BUNDLE
  }
```

Now I just need to figure out how to select/pin the poetry version renovate should use.

**Comment by zebulon501** [comment]
As a user, I shouldn't have to deal with how renovate and its external helper/tools are built. Defining a CA Bundle in a single way should be possible without the need of creating a new image,

**Comment by rarkins** [comment]
@zebulon501 contributions or constructive suggestions are welcome, complaints are not

**Comment by viceice** [comment]
the better solution is now using the slim image but set binary source to install. then no docker is required but renovate can still required tool versions on the fly.

**Comment by zebulon501** [comment]
Sorry, I didn't mean to be complaining, I simply wanted to rephrase the original post in a more concise manner with a user point of view without simply saying “I have the same need”.
I wanted to express the fact that renovate should work as a black box and forward the CA bundle to either sidecar containers (or locally installed tools, depending on the configuration). Needing to know in which language is written each and every tool used by renovate to be able to pass them the CA bundle in the correct way could be quite complex from a user perspective, especially if the need is to host it in a generic way for multiple teams.
By digging a bit more, it seems the right place for that would be https://github.com/renovatebot/docker-renovate, not directly this repo. I’ll take a deeper look at the various repos and try to come back with something more consistent.
Again, I didn’t mean to be aggressive, sorry for that.
